pkg/svc/zserver.go

```diff
@@ -2,51 +2,142 @@ package svc
-// versioning system, or get rid of it entirely. GetVersion helps make up a
-// simple "starter" example that allows an end-to-end example. It is not
-// required.
-//
-// TODO: Oh yeah, delete this guide when you no longer need it.
-//
-// ~~~~~~~~~~~~~~~~~~~~~~~~~~~ FAREWELL AND GOOD LUCK ~~~~~~~~~~~~~~~~~~~~~~~~~~
-// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-
 const (
 	// version is the current version of the service
 	version = "0.0.1"
 )
 
+var (
+	serverAddr = flag.String("addr", "localhost:9099", "The server address in the format of host:port")
+)
+
 // Default implementation of the MyApp server interface
-type server struct{}
+type server struct {
+	pb.UnimplementedMyAppServer
+	Description string
+	Timestamp   time.Time
+	Requests    int64
+}
+
+func GRPCConnect() (*grpc.ClientConn, error) {
+	var opts []grpc.DialOption
+	opts = append(opts, grpc.WithInsecure())
+	conn, err := grpc.Dial(*serverAddr, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
 
 // GetVersion returns the current version of the service
-func (server) GetVersion(context.Context, *empty.Empty) (*pb.VersionResponse, error) {
-	return &pb.VersionResponse{Version: version}, nil
+func (*server) GetVersion(context.Context, *empty.Empty) (*pb.VersionResponse, error) {
+	conn, err := GRPCConnect()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+	client := rpb.NewMyResponderClient(conn)
+	resp, err := client.GetVersion(context.Background(), &empty.Empty{})
+	if err != nil {
+		return nil, err
+	}
+	return &pb.VersionResponse{Version: resp.Version}, nil
+}
+func (s *server) UpdateDescription(ctx context.Context, req *pb.UpdateDescriptionRequest) (*pb.UpdateDescriptionResponse, error) {
+	if req.GetDescription() == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "Description can't be empty")
+	}
+	if req.GetService() != 1 {
+		conn, err := GRPCConnect()
+		if err != nil {
+			return nil, err
+		}
+		defer conn.Close()
+		client := rpb.NewMyResponderClient(conn)
+		resp, err := client.UpdateDescription(ctx, &rpb.UpdateDescriptionRequest{Description: req.Description, Service: req.Service})
+		if err != nil {
+			return nil, err
+		}
+		return &pb.UpdateDescriptionResponse{Description: resp.Description}, nil
+	}
+	s.Requests += 1
+	s.Description = req.Description
+	return &pb.UpdateDescriptionResponse{Description: s.Description}, nil
+}
+
+func (s *server) GetDescription(ctx context.Context, req *pb.GetDescriptionRequest) (*pb.GetDescriptionResponse, error) {
+	if req.GetService() != 1 {
+		conn, err := GRPCConnect()
+		if err != nil {
+			return nil, err
+		}
+		defer conn.Close()
+		client := rpb.NewMyResponderClient(conn)
+		resp, err := client.GetDescription(ctx, &rpb.GetDescriptionRequest{Service: req.Service})
+		if err != nil {
+			return nil, err
+		}
+		return &pb.GetDescriptionResponse{Description: resp.Description}, nil
+
+	}
+	s.Requests += 1
```

Comment:
1. Increment `s.Requests++` seems more preferable here.
2. Will this code work fine, processing a few requests simultaneously?

Reply:
1-Ok
2- No its not. I will fix this with mutexes

Reply:
Fixed